ARROW-5579: [Java] Shade flatbuffers

emkornfield opened the pull request:
- Merge the format package into the vector package
- shade flatbuffers in vector
- create missing wrappers to avoid direct flatbuffer dependencies in flight

[] Possibly add unit tests to cover new functions/classes.

@pravindra @tianchen92 please take a look, I think this is the only reasonable way to shade flatbuffers.  I don't know if there was any particular reason that format and vector where separate but I think  it makes it very difficult to shade if we don't take this approach.

tianchen92 commented:
Looks good @emkornfield, seems something wrong with the license?

emkornfield commented:
@pravindra mind taking a look?  Want to check if this works with Dremio code?

pravindra commented:
I will check if this works with dremio but I don't know enough about shading. 

@praveenbingo can you please review this one ?

wesm commented:
It seems like this is going to have to get fixed after 0.14.0. I don't think we should hold up the release over it

emkornfield commented:
@praveenbingo  mind taking a look?

elahrvivaz commented:
It might be better to create a third module that pulls in both `vector` and `format` and applies the shading there. I believe with the current changes in this PR, flatbuffers will still be a transitive dependency of the module, so importing projects will need to exclude it in their pom dependency, e.g.
```
        <dependency>
            <groupId>org.apache.arrow</groupId>
            <artifactId>arrow-vector</artifactId>
            <version>${arrow.version}</version>
            <scope>compile</scope>
            <exclusions>
                <exclusion>
                    <groupId>com.google.flatbuffers</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

With a third module, you could mark flatbuffers as `provided` scope so that it doesn't become a transitive dependency. Also, it would have the added benefit of not changing the current modules for users who don't care about the shading.

emkornfield commented:
@elahrvivaz I change to the approach you suggested.  I'm ambivalent if it is better.  It appears the shade plugin automatically moves shaded dependencies to provided scope, and with this approach we need to choose which module to rely on (shaded or unshaded), I chose all downstream packages to rely on the shaded version for safety/isolation, but consumers will also need to decide whether to use the shaded version or not.  Is there a good way to publicize which one to use.  It actually makes me a little nervous that people are directly relying on the format package and not going through the POJO wrappers.

elahrvivaz commented:
You could take the approach used by Apache Orc - create a secondary artifact that has the shading in it: https://github.com/apache/orc/blob/master/java/pom.xml#L275-L311. If you specify `<shadedArtifactAttached>true</shadedArtifactAttached>`, you will get two artifacts, one with a classifier of `shaded`.
Unfortunately, maven doesn't make this type of thing very easy, but I think having the unshaded jar available will be helpful for downstream projects that want to shade Arrow into a larger jar, or who already have the correct version of flatbuffers on the classpath.

elahrvivaz commented:
For publicizing the changes, maybe just add a note on the [install page](https://arrow.apache.org/install/) that jars are available shaded with flatbuffers or not.

emkornfield commented:
Are these theoretical concerns or do you have a project that breaks with the original shading approach?

elahrvivaz commented:
@emkornfield I work on [GeoMesa](https://github.com/locationtech/geomesa), we manage a lot of dependencies using maven (including Arrow), and use shading somewhat extensively. I think whatever approach you take, we will be able to work around, but from experience I would say that you never want to shade a module unless it's the 'final' jar being created (i.e. no other modules in the project depend on it). If you ever want to, for example, created a shaded jar in the `flight` module, you will run into issues if there is no unshaded `vector` jar (we tried something similar to shade Kryo versions and weren't able to get it working). For downstream projects, it's more an inconvenience (requiring the 'excludes' on import). But that is why most projects that provide shaded jars also provide non-shaded ones.
Anyway, just my $.02 - trying to save you some pain down the road :)

emkornfield commented:
@elahrvivaz thanks for the feedback, I'm fairly new to maven which is why I'm asking questions to determine the impact of various approaches.  A few more:
1.  Flight already shades some dependencies, would the fact that it still builds and passes unit tests with the shaded JAR indicate we've sufficiently isolated the flatbuffer dependencies?
2.  It seems like the approach of providing unshaded and shaded dependencies would require the same to be done for every module that arrow produces?  If this is the case would it be better to not shade and let clients worry about it?  
3.  I'm not sure if you looked at the original JIRA but the main reason for shading flatbuffers is that the generated flatbuffer files are only compatible with the same version of the flatbuffer library that generated them.  Does this change the considerations for shading?

elahrvivaz commented:
@emkornfield sure, happy to help. To your questions:
1. I'm not sure - the main issues I've seen related to trying to shade/relocate an already shaded/relocated dependency. Possibly being in the same project and having the same versions of flatbuffers is sufficient for that to work correctly. When I tried to do something similar a few years ago, I did not get it to work correctly, but it may have only broken my IDE (intellij) and not maven itself, or I may have been trying to use two different versions of the dependency (kryo) - I can't recall exactly what the problem was.
2. I think you are correct that each module would need a shaded and unshaded version. If you define the shade plugin in the parent pom, all the children will inherit it automatically. I think providing shaded artifacts can still be useful to clients, as I'll outline below.
3. I did see the ticket - shading seems a reasonable approach, but it might be worth checking out 'best practices' or how other flatbuffer projects handle the issue.

There are 3 main use cases I can think of, in increasing complexity:

1. In the general case, where a downstream project is either 1) using the same version of flatbuffers or 2) not using flatbuffers (and thus inheriting it from arrow), then shading is not necessary, but also not harmful. If the arrow jar is shaded, they may end up with flatbuffers on the classpath twice (once relocated and once not), since it is not removed from the dependency tree after shading. They could work around this by using an 'excludes' for flatbuffers in the arrow dependency declaration, or possibly using the 'dependency-reduced-pom' option with the shade plugin could fix this.
2. In a more advanced case, the project is trying to use an incompatible version of flatbuffers. In this case, they would need an intermediate artifact that shades and relocates flatbuffers into arrow - you can't manage two different versions of a dependency in a single pom, so you can't relocate one version of flatbuffers while also using a second version (in the same pom). If the arrow jar is not shaded, then they would need to create their own intermediate shaded artifact.
3. Finally, the project might need to shade and relocate arrow itself. I think that this would require an non-shaded arrow artifact to work with, as I don't think the shade plugin can handle relocating classes twice.

I think the most flexible approach would be to declare the shade plugin in the parent pom, using `<shadedArtifactAttached>true</shadedArtifactAttached>`. That will create both shaded and non-shaded jars for each module, which should cover all the use cases. I think that you could also define the flatbuffers dependency as a 'hard' dependency by using [bracket notation](https://maven.apache.org/pom.html#Dependency_Version_Requirement_Specification). That would force downstream projects to reconcile the flatbuffers version in their pom (I haven't ever tried using a hard version though, so it might not work exactly like that...). For most projects, they would continue using the regular artifacts. If they try to use an incompatible flatbuffers version, maven should fail the build, at which point they can switch to the shaded classifier and add the excludes for flatbuffers.

BryanCutler commented:
I'm not an expert on shading, but @elahrvivaz proposal above sounds like the best approach to me. Correct me if I'm wrong, but this will not require any code changes because the shade plugin handles all translations right?

emkornfield commented:
@elahrvivaz @BryanCutler thanks for all the input.  I've updated the approach to mostly align with the suggestions, but with some small changes:

1.  I provide an additional shaded JAR containing flatbuffers and format directly in vector.  The main downside to this versus a separate package is we can't exclude the dependencies ourselves, but I think it is simpler and should avoid maven analysis failures.  More advanced consumers should be able to do it easily add the exclusion if necessary.
2.  I added a classifier option to all packages that depend on arrow-vector (I'm still pruning all dependencies on format outside of vector) which can be set with a profile, so we can run tests with shading.  I've attempt to test the profile in the CI docker image for java 11.
3.  I'd rather not experiment with the hard dependency flag without a way to test it.  Maybe a follow-up JIRA if either one of you want to play with it?

One thing that makes this more complicated is that the flatbuffer generated files extend classes from the  runtime library.  Prior wrapping all of the generated code from arrow-format in arrow-vector,  shading  wasn't distributive. ShadeFlatbuffers(arrow-flight, arrow-vector, arrow-format) will work but needs to have all packages shaded and included in the uber jar ShadeFlatbuffers(arrow-flight) + ShadeFlatbuffers(arrow-format) + ShadeFlatbuffers(arrow-vector) would break.

Regarding recommendations from the flatbuffer community, if you follow the linked thread not much guidance was given so we are on our won.

emkornfield commented:
@elahrvivaz @BryanCutler @praveenbingo would one of you mind reviewing?

BryanCutler commented on java/flight/pom.xml:
```diff
@@ -25,25 +25,17 @@
       <groupId>org.apache.arrow</groupId>
       <artifactId>arrow-vector</artifactId>
       <version>${project.version}</version>
+      <classifier>${arrow.vector.classifier}</classifier>
```
Is this necessary? The shade-plugin can already add a classifier name

emkornfield replied:
I'm not sure I understand the question.  This is intended to allow building/running tests with either the shaded or non-shaded artifact.

BryanCutler replied:
I misunderstood the purpose of this, seems fine

BryanCutler commented on java/pom.xml:
```diff
@@ -646,8 +642,8 @@
   <modules>
-    <module>format</module>
     <module>memory</module>
+    <module>format</module>
```
Was there a reason to switch these?

emkornfield replied:
No, changed it back.

BryanCutler commented on java/pom.xml:
```diff
@@ -679,6 +675,14 @@
       </modules>
     </profile>
+
+    <profile>
```
Maybe I'm missing something but I thought that the shaded artifacts are always produced alongside the normal ones? So I'm not sure why we need a profile

emkornfield replied:
My intent was to allow testing the the shaded artifact with downstream modules.  (see the changes to the ci shell scripts where I tried to excercise this profile).

BryanCutler replied:
I get it now, seems fine to me

emkornfield commented:
@elahrvivaz I think the changes to flatbuffer code are still needed if we want to produce a limited subset of shaded artifacts.  See note above about non-distributiveness of shading.  TL;DR; if we don't make the code changes we need to add shading to all modules that depend on arrow-vector and make use of flat-buffer code directly, this shading would need to include the arrow-vector and arrow-flatbuffer modules directly in the packaged JARs as well.  This seems like a longer term maintenance burden.

@BryanCutler for dividing up the changes where you thinking of a separate PR or rebase with two commits, one containing the code and one containing everything else?

emkornfield commented:
@BryanCutler I've squashed the changes into 2 commits.  The first contains only java code changes, the second contains POM + auxiliary changes.

BryanCutler commented:
Thanks for squashing the commits @emkornfield , I will definitely take another closer look at this in the next few days.

emkornfield commented:
@BryanCutler did you get a chance to look at this again?

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java:
```diff
@@ -106,6 +106,16 @@ private ArrowRecordBatch(boolean dummy, int length, List<ArrowFieldNode> nodes,
+  /**
+   * Returns the serialized form of {@link RecordBatch} wrapped in a {@link org.apache.arrow.flatbuf.Message}.
+   */
+  public ByteBuffer toSerializedFlatBuffer() {
```
I'm not sure about adding this API. I don't think it is something the end user will call, and it's a little confusing that it's only the metadata. This is only used in Flight correct? Could this be added to `MessageSerializer` instead or change `MessageSerializer.serialize()` to create a `FlatBufferBuilder` itself?

emkornfield replied:
moving to serializer

emkornfield replied:
I also moved all the other toSerializedFlatbuffer additions to MessageSerializer (will push new commit shortly).

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Schema.java:
```diff
@@ -79,6 +82,20 @@ public static Schema deserialize(ByteBuffer buffer) {
+   * <p>The message must have the schema field set or an
+   * exception will be raised.
+   */
+  public static Schema deserializeFromMessage(ArrowBatchMessage buffer) {
```
Not sure about this one either, what's the difference between this and `MessageSerializer.deserializeSchema()`?

emkornfield replied:
it appears it might not be needed.

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -576,4 +601,5 @@ public static ArrowBuf readMessageBody(ReadChannel in, int bodyLength, BufferAll
     }
     return bodyBuffer;
   }
+
```
was this intentional?

emkornfield replied:
no, or at least its been long enough that I don't remember.

BryanCutler commented on java/flight/src/main/java/org/apache/arrow/flight/ArrowMessage.java:
```diff
@@ -182,27 +169,20 @@ public HeaderType getMessageType() {
     return HeaderType.getHeader(message.headerType());
   }
 
-  public Message asSchemaMessage() {
```
So this needs to be removed since it returns a FlatBuffer message, but would we need an alternative that returns the wrapper so someone could still get the FlatBuffer message?

emkornfield replied:
Potentially, I think given this refactoring that fact the no-one else needs the reference means we can expose it at a later time.

BryanCutler replied:
Since it's a public API, I'm a little worried that someone might depend on this downstream. If we had
`public MessageMetadataResult asSchemaMessage()` then someone could still get the `Message` by calling `MessageMetadataResult.getMessage()` if needed right?

emkornfield replied:
added back.  I think in general we might be exposing too many public APIs without cause, but it is better to not make clients wait 2 release cycles if someone does in fact depend on this.

BryanCutler replied:
>I think in general we might be exposing too many public APIs without cause

I agree. I'm not sure how this is used, but since it's already exposed publicly, we should be careful about changes.

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowBatchMessage.java:
```diff
@@ -0,0 +1,52 @@
+ * Wrapper around flatbuffer message table.  This keeps from exposing
+ * flatbuffer details to upstream packages.
+ */
+public class ArrowBatchMessage {
```
is the name `ArrowBatchMessage` misleading because this can wrap all `Message`s not just `ArrowRecordBatch` or `ArrowDictionaryBatch` right?

emkornfield replied:
renamed to ArrowMessageHeader

emkornfield replied:
Actually that is always something in the flatbuffers schema.  let me pick something different.

emkornfield replied:
renamed to ArrowMessageMetadata

emkornfield replied:
It turns out this duplicate most of the functionality on an already existing class.  I've removed this and added functionality to the new class.

BryanCutler commented on java/pom.xml:
```diff
@@ -486,11 +487,6 @@
-        <groupId>com.google.flatbuffers</groupId>
-        <artifactId>flatbuffers-java</artifactId>
-        <version>${dep.fbs.version}</version>
-      </dependency>
```
Since this is in the dependency management section, couldn't it stay here?

emkornfield replied:
yes, I misunderstood what this section does.

BryanCutler commented on java/format/pom.xml:
```diff
@@ -33,6 +33,7 @@
     <dependency>
       <groupId>com.google.flatbuffers</groupId>
       <artifactId>flatbuffers-java</artifactId>
+      <version>${dep.fbs.version}</version>
```
I think if we leave flatbuffers in the dependency management section, then this could be removed also

emkornfield replied:
yes.  done.

BryanCutler commented on java/vector/pom.xml:
```diff
@@ -133,6 +134,37 @@
+      <plugin>
+       <groupId>org.apache.maven.plugins</groupId>
+       <artifactId>maven-shade-plugin</artifactId>
+       <version>3.2.1</version>
```
Can we stick to the same version of shade plugin? it looks like 3.1.1 in flight

emkornfield replied:
yes.

BryanCutler commented on java/vector/pom.xml:
```diff
@@ -133,6 +134,37 @@
+             <relocations>
+               <relocation>
+                 <pattern>com.google.flatbuffers</pattern>
+                 <shadedPattern>org.apache.arrow.shaded.com.google.flatbuffers</shadedPattern>
```
following the shadedPattern in flight, it would be `arrow.vector.com.google.flatbuffers`, could we stick to this format?

emkornfield replied:
done.

emkornfield commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -340,6 +343,14 @@ public static ArrowRecordBatch deserializeRecordBatch(RecordBatch recordBatchFB,
   }
 
+  /**
+   * Reads a record batch from the byte bufer and underlying data buffer.
```
rephrased.

emkornfield commented:
@BryanCutler thanks for the review, I believe I addressed all of your comments.

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageMetadataResult.java:
```diff
@@ -42,25 +42,39 @@
+   *
+   * @deprecated unused.
    */
+  @Deprecated
```
I use this downstream in Spark, can you not deprecate it? I found it useful to not have to rely on calculating the ByteBuffer size that is created in Arrow when reading from a `ReadChannel`.

BryanCutler replied:
great, thanks!

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageMetadataResult.java:
```diff
@@ -42,25 +42,39 @@
     this.message = message;
   }
 
+  public static MessageMetadataResult create(ByteBuffer buffer) {
```
Can you add a javadoc?

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageMetadataResult.java:
```diff
@@ -42,25 +42,39 @@
   }
 
+  public static MessageMetadataResult create(ByteBuffer buffer) {
+    return new MessageMetadataResult(-1, buffer, Message.getRootAsMessage(buffer));
```
instead of -1 could you calculate a length from `buffer`?

emkornfield replied:
I added it as an explicit parameter.

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageMetadataResult.java:
```diff
@@ -42,25 +42,39 @@
     return messageBuffer;
   }
 
+  public int bytesAfterMessage() {
```
is this meant to be the message body length?

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -380,6 +403,16 @@ public static ArrowBlock serialize(WriteChannel out, ArrowDictionaryBatch batch)
   }
 
+  /** Returns the serialized flatbuffer bytes for this object. */
+  public static ByteBuffer serializeMetadata(ArrowDictionaryBatch batch) {
```
could you make one generic `serializeMetadata(ArrowMessage message, byte headerType)` instead of 3 different ones?

emkornfield replied:
Schema doesn't implement ArrowMessage and I'd prefer not to make the change necessary to do it.  I have consolidated the two batch types by adding a getType to ArrowMessage interface.

BryanCutler replied:
ok, sounds fine

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -340,6 +357,15 @@ public static ArrowRecordBatch deserializeRecordBatch(RecordBatch recordBatchFB,
+  /**
+   * Reads a record batch based on the metadata in serializedMessage and the underlying data buffer.
+   */
+  public static ArrowRecordBatch deserializeRecordBatch(MessageMetadataResult serializedMessage,
```
is it necessary to add these that just wrap the call to `getMessage()`?

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Schema.java:
```diff
@@ -79,6 +81,20 @@ public static Schema deserialize(ByteBuffer buffer) {
+   * <p>The message must have the schema field set or an
+   * exception will be raised.
+   */
+  public static Schema deserializeFromMessage(MessageMetadataResult buffer) {
```
This is duplicated from MessageSerializer right? I saw you removed `deserialzeFromMessage` for the other messages, can you do the same here?

emkornfield replied:
I think I copied it from flight originally, but removed.

BryanCutler commented:
Thanks @emkornfield , I will take another pass on Monday. Hopefully we can get some more eyes on this by then too.

emkornfield commented:
@BryanCutler since no one else has expressed an opinion, what do you think about going ahead and merging this?

praveenbingo commented:
@emkornfield sorry for the delay, was in a long hiatus.

I am trying this out with dremio, will ping back if i run into any issues by end of day. 
Also looks like this needs a rebase with the master.

BryanCutler commented on ci/docker_java_test_all.sh:
```diff
@@ -28,6 +28,7 @@ export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
 export ARROW_JAVADOC=1
 bash $SOURCE_DIR/docker_build_java.sh
 
+export ARROW_JAVA_SHADE_FLATBUFS=1
```
Does this mean that CI is only going to test with the shaded artifacts?

emkornfield replied:
Java 8 is tested unshaded java-11 is tested with shaded.

emkornfield replied:
(and I think this only applies to UrsaBot runs).

BryanCutler replied:
Ah right, ok sounds good

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowMessage.java:
```diff
@@ -26,6 +26,8 @@
 
   <T> T accepts(ArrowMessageVisitor<T> visitor);
 
+  byte getType();
```
`getType` is a little ambiguous. How about `getHeaderType`?

emkornfield replied:
renamed to something clearer and added comment.

emkornfield commented:
@praveenbingo @BryanCutler rebased.  I think rerunning the one fail test because it seems due to network flakiness.

praveenbingo commented:
@emkornfield please go ahead with the change.

I am having a lot of trouble to push this through our pipeline (nothing related to the change itself). I will pick up any changes if required in another patch.

BryanCutler commented:
Thanks @praveenbingo , I will go ahead and merge. We can address any issues as followups.

BryanCutler commented:
merged to master, thanks @emkornfield !
